### ISSUE TITLE
[render issue] switch brightness causes wrong partial render.

### ISSUE THREAD

**Issue by joechan-cq** [flutter/flutter#129000]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. open a flutter page.
2. swtich brightness on system.

### Expected results

the page rendered to correct color.

### Actual results

1. part of page rendered. the rest was not updated.
2. the page was flash if scrolling the listview, it looks like keeping switch the dark and light color.

### Code sample

<details><summary>Code sample</summary>

```dart
import 'dart:ui';

import 'package:flutter/material.dart';

class TestList extends StatefulWidget {
  const TestList({Key? key}) : super(key: key);

  @override
  State<TestList> createState() => _TestListState();
}

class _TestListState extends State<TestList> with WidgetsBindingObserver {
  late Brightness brightness;

  @override
  void initState() {
    WidgetsBinding.instance.addObserver(this);
    brightness = window.platformBrightness;
    super.initState();
  }

  @mustCallSuper
  @override
  void didChangePlatformBrightness() {
    brightness = window.platformBrightness;
    if (mounted) {
      setState(() {});
    }
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('List Test Page'),
      ),
      body: SafeArea(
        child: Container(
          color: brightness == Brightness.light ? Colors.white : Colors.black,
          child: ListView.builder(
            itemCount: 100,
            itemBuilder: (ctx, index) {
              return Container(
                padding: EdgeInsets.all(20),
                child: Text(
                  '$index',
                  style: TextStyle(
                    fontSize: 18,
                    color: brightness == Brightness.light ? Colors.black : Colors.white,
                  ),
                ),
              );
            },
          ),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/9456005/73810b85-24f3-4370-9976-da8b8e25aea7

![Screenshot_20230616_150609](https://github.com/flutter/flutter/assets/9456005/f18f4664-b6a3-44a5-ae9c-213291faafe6)

</details>

### Logs

It looks like that the render issue results from different refreshing frequencies between several areas of the picture. 
The issue only happens when I integrate the program with a hybrid development project (with Flutter_Boost). I cannot reproduce the problem when running the program independently. I am not sure if this is caused by the Native code of the hybrid project or if it is an issue with Flutter itself.
The issue only appears when I try to upgrade the SDK from 2.5.3 to 3.3.10 or 3.7.11. Previously it works perfectly with SDK 2.5.3
I found some possible related issues: #111376, #110885。

My device：xiaomi mix2 and Huawei mate 50 pro.

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.3.10, on macOS 13.2.1 22D68 darwin-arm, locale zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
[!] Xcode - develop for iOS and macOS (Xcode 14.2)
    ! CocoaPods 1.10.1 out of date (1.11.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.3)
[✓] VS Code (version 1.78.1)
[✓] Connected device (3 available)
[✓] HTTP Host Availability

! Doctor found issues in 1 category.

```

</details>

**Comment by dam-ease** [comment]
Hi @joechan-cq. I see you aren't on the latest flutter version. Kindly confirm this issue still persists by upgrading to the latest `stable` version.

**Comment by joechan-cq** [comment]
I tried it on Flutter 3.10.5, and it worked fine. So I will close this issue although I can't use the newest flutter sdk for some other reasons.

I would appreciate it if I could acquire the way to fix the issue on Flutter 3.3.10.
